### ISSUE TITLE
Speed up tests by re-using the database

### ISSUE THREAD

**Issue by ewjoachim** [peopledoc/procrastinate#238]
Currently, each acceptance and integration tests use a new database. We use "CREATE DATABASE FROM template", but there is still a non-negligeable time used to setup an entirely new database for each test.

I *think* (without being 100% sure) a good number of tests would behave just fine if we ran them in a re-used database, in a transaction that we would rollback afterwards.

We should try having a new fixture that would reuse the same DB and rollback, and see.

An alternative strategy would a fixture that empties our 2 tables before and/or after each test, using either `DELETE` or `TRUNCATE` (though... I think I recall `TRUNCATE` can be slower in some cases).

Also, any kind of DB optimization disabling postgres "durability" features in tests is welcome in that PR.

**Comment by ewjoachim** [comment]
https://docs.docker.com/registry/storage-drivers/inmemory/ ?

**Comment by ewjoachim** [comment]
postgres.conf:
```
  fsync = off                # turns forced synchronization on or off
  synchronous_commit = off   # synchronization level; on, off, or local
  full_page_writes = off     # recover from partial page writes
```

**Comment by ewjoachim** [comment]
I've note it as both "Good for: newcomers" and "Good for: Occasional contributors", depending on how far you're willing to go :)

Adjusting the docker-compose.yaml for better performance and not touching the code is a perfectly valid contribution !

**Comment by elemoine** [comment]
> I _think_ (without being 100% sure) a good number of tests would behave just fine if we ran them in a re-used database, in a transaction that we would rollback afterwards.

The problem is that we cannot rollback the transaction. Our aiopg connector uses this to issue queries to the database:

https://github.com/peopledoc/procrastinate/blob/77e036337441129f48a021bc5b8e6a8978aa3a78/procrastinate/aiopg_connector.py#L249-L252

`with await self.pool.cursor()` acquires a connection from the pool and creates a cursor on that connection. The connection works in `autocommit` mode, meaning that the query sent to Postgres has immediate effect.

Bottom line is that we cannot have control on the connection (and transaction) our aiopg code uses for issuing queries to the database system.

**Comment by elemoine** [comment]
I am currently looking at the `TRUNCATE` option…

**Comment by elemoine** [comment]
> I am currently looking at the `TRUNCATE` option…

With https://github.com/peopledoc/procrastinate/commit/49f0098bb610428ca2a4d7fabd13dc05af0541cb running all the integration tests takes ~7 seconds on my machine, while it takes ~34 seconds with the current master branch. So a 5x speed-up.

**Comment by elemoine** [comment]
With https://github.com/peopledoc/procrastinate/pull/315, and the use of Postgres 12, we made our test suite run in ~30 seconds rather than ~1 minute. And with this, the use of `TRUNCATE` and `ALTER SEQUENCE` to kind of reset the database (https://github.com/peopledoc/procrastinate/commit/46d2b22ab84f8537412c92c929e156918e279ff8) does no longer significantly speed up the execution of the test suite. I can observe a small improvement, but I don't think it justifies switching to a less robust mechanism.

**Comment by ewjoachim** [comment]
So, shall we close ?

**Comment by elemoine** [comment]
> So, shall we close ?

Yep, I think we can close.
